### PR TITLE
Channel Import Script

### PR DESCRIPTION
**Pull request by Murodese**
⚠️ please review data import PR first https://github.com/breadtubetv/breadtubetv/pull/87 ⚠️ 

## Development

You will need a clone of the project in a specific directory, you will need [Go set up on your machine](https://golang.org/doc/install).

```
cd ~/.go/src/github.com/
git clone git@github.com:breadtubetv/breadtubetv.git
cd ~/.go/src/github.com/breadtubetv/bake
go get
go run main.go config youtube #follow prompts
go run main.go channel import petercoffin youtube https://www.youtube.com/user/petercoffin
```

## Requirements

- [x] Connect to YouTube API [@Murodese]
- [x] Retrieve Channel Information from YouTube API [@Murodese]
- [x] Store YouTube information in Provider Namespace https://github.com/breadtubetv/breadtubetv/pull/54#issue-247587101 [@j16r]
  - [x] Store meta information on first create
  - [x] Do not overwrite outside of providers context
- [x] Utilize channel slug to store channel in specified file https://github.com/breadtubetv/breadtubetv/pull/54#issue-247587101 [@j16r]

## New Stories

- [ ] Calculate the subscriber count as a count of all provider subscribers
- [ ] A command to update all channels, eg `bake channel update` [@jaymickey]

## Explanation

This CLI forms the basis of future data management efforts, allowing non-technical users to add channels to Breadtube. Designed to be provider-agnostic (currently just Youtube).

Intended to be used as:

`bake config [provider]`
`bake channel import [slug] [provider] [url]`

eg:

`bake channel import example youtube https://www.youtube.com/channel/123cde456efg`
`bake channel import petercoffin youtube https://www.youtube.com/user/petercoffin`

Later:

`bake video import [provider] [url]`, etc.

Available providers:
- Youtube (`youtube`)

Completes #37.

### CODE REVIEW

**Comment by Murodese** [comment]
Ready for review @dirkkelly. 

Test case:
```bash
> cd bake
> go run main.go config youtube
> go run main.go channel import youtube https://www.youtube.com/channel/UCL2dVYeD8YGfMWPXDV-QjUQ
(run it twice to check)
> go run main.go channel import youtube https://www.youtube.com/channel/UCL2dVYeD8YGfMWPXDV-QjUQ
```

Did we want unit tests for this?

**Comment by Murodese** [comment]
Whoops, just noticed you split the ymls. Any reason why?

**Comment by dirkkelly** [comment]
@Murodese split the yml files so that people didn't have to maintain array syntax and order of subscribers. Sorry I  texted but obvi too late.

File: bake/cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/breadtubetv/bake/providers"
```
Comment (dirkkelly) [inline bake/cmd/root.go]: should this be `github.com/breadtubetv/breadtubetv/bake/providers`?

File: bake/cmd/import.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
+// Copyright © 2019 NAME HERE <EMAIL ADDRESS>
```
Comment (dirkkelly) [inline bake/cmd/import.go]: current license on project is GPLv3, does this conflict?

File: bake/cmd/root.go (hunk 1 above)
Comment (Murodese) [reply bake/cmd/root.go]: Nah, the go formatter in VS Code automatically resolves the package refs so they're all correct.

File: bake/cmd/import.go (hunk 2 above)
Comment (Murodese) [reply bake/cmd/import.go]: Yeah, I'll remove it. The generator automatically APL's everything and I forgot it from a few files.

**Comment by dirkkelly** [comment]
## Successs!

```
cd ~/.go/src/github.com/
git clone git@github.com:breadtubetv/breadtubetv.git
cd ~/.go/src/github.com/breadtubetv/bake
go get
go run main.go config youtube
go run main.go channel import youtube https://www.youtube.com/channel/UCEWrkBTDSYqJzQmp8gxQiMw
```

**Comment by dirkkelly** [comment]
- Can we protect existing attributes like description? 
- Can we put all attributes in double quotes to protect against weird characters?

screen shot 2019-01-25 at 11 08 39 am

**Comment by Murodese** [comment]
> * Can we protect existing attributes like description?

Only if they're "standard". Unfortunately go's statically-typed, so you have to specify the yaml structure to marshal/unmarshal it. The easy way to do that I guess would be to have a map field (which gets unmarshalled to a dictionary).

> * Can we put all attributes in double quotes to protect against weird characters?

Yeah, I'll see if I can change the rules it's using.

**Comment by Murodese** [comment]
Go-yaml autoquotes strings that require it, but don't have an option to manually mark all fields as quoted (...).

**Comment by j16r** [comment]
Finished main requirements I believe :bowtie: 

Notes: The yaml parser we're using doesn't have a great way to compose, so the amount of yaml parsing is a bit excessive in order to handle leaving around unrecognized keys in the top level document (remnants as I've called them).

Also added testify as it has better matchers and actual line numbers upon failure, used in small tests. Note the tests will reach out to YouTube right now.

**Comment by dirkkelly** [comment]
Working with @j16r to figure this out

`Error obtaining channel info: Could not find channel from URL.`

Which is caused when sending through a user url instead.

```
go run main.go channel import youtube perseus999 https://www.youtube.com/user/Perseus999
go run main.go channel import youtube perseus999 https://www.youtube.com/Perseus999
```

Also @j16r can you please change the argument order to swap slug and provider

```
bake channel import [slug] [provider] [channel_url] [flags]
```

**Comment by dirkkelly** [comment]
- [ ] @j16r I'm losing the tags data when I update a previously existing channel

original

```yaml
name: "Sarcasmitron"
slug: "sarcasmitron"
url: "https://www.youtube.com/channel/UC6WD1VImCVeOuNccuqDzAdA"
subscribers: 2040
tags: ["breadtube"]
```

updated 

```
name: Sarcasmitron
providers:
  youtube:
    name: Sarcasmitron
    slug: UC6WD1VImCVeOuNccuqDzAdA
    url: https://www.youtube.com/channel/UC6WD1VImCVeOuNccuqDzAdA
    description: ""
    subscribers: 2157
slug: sarcasmitron
```

**Comment by dirkkelly** [comment]
```
name: "potholer54"
slug: "potholer54"
url: "https://www.youtube.com/channel/UCljE1ODdSF7LS9xx9eWq0GQ"
description: "The main purpose of this channel is to explain in simple terms the conclusions of scientific research, and correct some of the unsourced crap we get fed on the Internet."
subscribers: 175241
tags: ["breadtube"]
```

```
name: "potholer54"
providers:
  youtube:
    name: potholer54
    slug: UCljE1ODdSF7LS9xx9eWq0GQ
    url: https://www.youtube.com/channel/UCljE1ODdSF7LS9xx9eWq0GQ
    description: ""
    subscribers: 176844
slug: potholer54
tags: ["breadtube"]
```

File: README.md [hunk 3]
```diff
@@ -43,6 +43,66 @@ open localhost:1313
...
+#### Installing Go
+
+This is going to be dependent on your system, we recommend following https://golang.org/doc/install
+
+#### Developing
+
+You'll need a copy of the project in the relevent folder.
+
+```
+mkdir -p ~/.go/src/github.com/
```
Comment (jaymickey) [inline README.md]: Developers should be able to use `go get github.com/breadtubetv/breadtubetv` which will create the folder structure and clone the project. They'll just have to `cd $GOPATH/src/github.com/breadtubetv/breadtubetv` and checkout the relevant branch.

**Comment by jaymickey** [comment]
Is the OP up to date with the development work still to be done?

> * Running update without arguments updates all

Should we create a `bake channels update` command?

File: README.md (hunk 3 above)
Comment (dirkkelly) [reply README.md]: There is currently an issue with the library details not matching the actual endpoint of the project.

**Comment by dirkkelly** [comment]
@jaymickey just updated the OP, looks like this branch is working though so I'd like to merge and update all the data. I think the stretch goals should be handled in further pull requests.

**Pull request merged**